File: paython/dijkstra.py
class Graph:
    def __init__(self) -> None:
        self.nodes = []
        self.edges = {}
        
    def add_node(self, node):
        self.nodes.append(node)
        self.edges[node] = []
    
    def add_edge(self, node1, node2, weight):
        self.edges[node1].append({"node": node2, "weight": weight})
        self.edges[node2].append({"node": node1, "weight": weight})
    
    def dijkstra(self, _from, _to):
        distances = {}
        parents = {}
        visited = []
        
        for node in self.nodes:
            distances[node] = float("inf")
            parents[node] = None
        distances[_from] = 0
        
        node = self._find_smollest_node(distances, visited)
        while node is not None:
            print(node)
            if node == _to:
                path = []
                at = _to
                while at != None:
                    path.insert(0, at)
                    at = parents[at]
                return {"distance": distances[_to], "path": path}
            
            for neighbor in self.edges[node]:
                alt = distances[node] + neighbor["weight"]
                if alt < distances[neighbor["node"]]:
                    distances[neighbor["node"]] = alt
                    parents[neighbor["node"]] = node
            visited.append(node)
            node = self._find_smollest_node(distances, visited)
            
    def _find_smollest_node(self, distances, visited):
        print("in smpll####")
        print(distances)
        print(visited)
        smollest_cost = float("inf")
        smollest_cost_node = None
        for node in distances:
            cost = distances[node]
            if cost < smollest_cost and node not in visited:
                smollest_cost = cost
                smollest_cost_node = node
                
        return smollest_cost_node
    
    
graph = Graph()
graph.add_node("A")
graph.add_node("B")
graph.add_node("C")
graph.add_node("D")

graph.add_edge("A", "B", 6)
graph.add_edge("A", "C", 2)
graph.add_edge("B", "D", 1)
graph.add_edge("C", "B", 3)

File: js/dijkstra.js
class Graph{
    constructor(){
        this.nodes = [];
        this.edges = {};
    }

    addNode(node){
        this.nodes.push(node)
        this.edges[node] = []
    }

    addEdge(node1, node2, weight){
        this.edges[node1].push({"node": node2, weight})
        this.edges[node2].push({"node": node1, weight})
    }

    dijkstra(start, end){
        let distances = new Map();
        let parents = new Map();
        let visited = new Set();

        this.nodes.forEach(node => {
            distances.set(node, node == start ? 0 : Infinity);
            parents.set(node, null);
        })

        let node = this._findSmollestNode(distances, visited);
        while(node != null){
            if(node === end){
                let path = [];
                let at = end;

                while(at != null){
                    path.unshift(at);
                    at = parents.get(at);
                }

                return {"distance": distances.get(end), "path": path};
            }

            this.edges[node].forEach(beighbor => {
                let cost = distances.get(node) + beighbor.weight;
                if(cost < distances.get(beighbor.node)){
                    distances.set(beighbor.node,cost);
                    parents.set(beighbor.node, node);
                }
            })
            visited.add(node);
            node = this._findSmollestNode(distances, visited);
        }
    }

    _findSmollestNode(distances, visited){
        let smollestCost = Infinity;
        let smollestCostNode = null;

        distances.forEach((distance, key) => {
            if(smollestCost > distance && !visited.has(key)){
                smollestCost = distance;
                smollestCostNode = key;
            }
        })
        return smollestCostNode;
    }
}

let graph = new Graph();
graph.addNode("A");
graph.addNode("B");
graph.addNode("C");
graph.addNode("D");
graph.addNode("E");
graph.addNode("F");

graph.addEdge("A", "B", 5);
graph.addEdge("A", "C", 2);
graph.addEdge("C", "B", 8);
graph.addEdge("C", "E", 7);
graph.addEdge("B", "E", 2);
graph.addEdge("B", "D", 4);
graph.addEdge("D", "E", 6);
graph.addEdge("D", "F", 3);
graph.addEdge("E", "F", 1);

console.log(graph)

let result = graph.dijkstra("A", "F");

console.log("###########  ", result)